File: examples/sub_bind.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%%! -smp enable -sname sub_bind -pa ../_rel/treehouse_release/lib/stdlib-3.0/ebin ../_rel/treehouse_release/lib/kernel-5.0/ebin ../_rel/treehouse_release/lib/chumak-1.1.1/ebin ../_rel/treehouse_release/lib/uuid-1.5.2-rc1/ebin ../_rel/treehouse_release/lib/luerl-0.3/ebin

main(_) ->
	io:format("Subscriber bind treehouse OTP release erlang escript.\n",[]),
    application:ensure_all_started(chumak),
    {ok, Socket} = chumak:socket(sub),

    %% List of topics, put them on a list!
    Topic = <<" ">>,
    Heartbeat = <<"heartbeat">>,
    Asterisk = <<"asterisk">>,
    Currency = <<"currency">>,
    Beam = <<"beam">>,
    Logging = <<"logging">>,
    Upload = <<"upload">>,
    %% Erlang zmq subscribe socket and topics!
    chumak:subscribe(Socket, Topic),
    chumak:subscribe(Socket, Heartbeat),
    chumak:subscribe(Socket, Asterisk),
    chumak:subscribe(Socket, Currency),
    chumak:subscribe(Socket, Logging),
    chumak:subscribe(Socket, Upload),
    
    case chumak:bind(Socket, tcp, "localhost", 8135) of
        {ok, _BindPid} ->
            io:format("Binding OK with Pid: ~p\n", [Socket]);
        {error, Reason} ->
            io:format("Connection Failed for this reason: ~p\n", [Reason]);
        X ->
            io:format("Unhandled reply for bind ~p \n", [X])
    end,
    loop(Socket).

loop(Socket) ->
    {ok, Data1} = chumak:recv(Socket),
    io:format("Received ~p\n", [Data1]),
    loop(Socket),
    erlang:halt(0).

File: examples/sub_bind.py
#!/usr/bin/env python

import logging, uuid, zmq

from zmq.eventloop import ioloop
from zmq.eventloop.future import Context, Poller

from tornado import httpclient as _http_client
from tornado import gen
import ujson as json

ioloop.install()

@gen.coroutine
def subscriber(port=5813):
    '''
        Bind Subscriber
    '''
    logging.warning("Binding SUB socket on port: {0}".format(port))
    context = Context()
    sub = context.socket(zmq.SUB)
    sub.bind("tcp://*:%s" % port)

    sub.setsockopt(zmq.SUBSCRIBE, " ")
    sub.setsockopt(zmq.SUBSCRIBE, "heartbeat")
    sub.setsockopt(zmq.SUBSCRIBE, "asterisk")
    sub.setsockopt(zmq.SUBSCRIBE, "logging")
    sub.setsockopt(zmq.SUBSCRIBE, "upload")
    sub.setsockopt(zmq.SUBSCRIBE, "beam")

    poller = Poller()
    poller.register(sub, zmq.POLLIN)

    http_client = _http_client.AsyncHTTPClient()

    while True:
        events = yield poller.poll(timeout=1000)
        if sub in dict(events):
            # receive raw msg from sub
            msg = yield sub.recv()
            # get topic and message
            topic = msg.split(' ')[0]
            message = ' '.join(msg.split(' ')[1:])
            # this make more sense directly on the beam
            # that why we're moving, still this sub_bind.py
            # clear some ideas directly.
            if topic.startswith('heartbeat'):
                print(topic, message)
            elif topic.startswith('asterisk'):
                print(topic, message)
            elif topic.startswith('logging'):
                print(topic, message)
            elif topic.startswith('upload'):
                print(topic, message)
            elif topic.startswith('beam'):
                print(topic, message)
            else:
                # let it crash
                logging.warning('let it crash')
                print(msg)
        else:
            #logging.warning('nothing receive')
            pass

def main():
    '''
        main function
    '''
    loop = ioloop.IOLoop.instance()
    loop.add_callback(subscriber)
    loop.start()

if __name__ == '__main__':
    main()